File: 백준/Silver/5525. IOIOI/IOIOI.py
import sys
input = sys.stdin.readline

N = int(input())
M = int(input())
S = input().rstrip()
P = 'IOI' + 'OI' * (N-1)
K = [0] * len(P)

def pattern (S):

    i = 0
    j = 1
    while j < len(S):
        if S[i] == S[j]:
            K[j] = i + 1
            i += 1
            j += 1
        else:
            j += 1

pattern(P)
cnt = 0

def KMP(P, S):

    global cnt
    i = 0
    j = 0
    while i < len(S):

        if S[i] == P[j]:
            if j == len(P) - 1:
                cnt += 1
                j = K[j-1]
            else:
                i += 1
                j += 1

        else:
            if j > 0:
                j = K[j-1]
            else:
                i += 1

KMP(P, S)
print(cnt)

File: 백준/Silver/5525. IOIOI/IOIOI.js
let fs = require('fs')
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n")
// let input = fs.readFileSync("./EX.txt").toString().trim().split("\n")

const N = parseInt(input[0])
const M = parseInt(input[1])
const S = input[2]

const len = 2 * N + 1

let i = 0
let cnt = 0
let ans = 0

while (i < M){

    x = S[i] + S[i+1] + S[i+2]

    if (x === "IOI") {
        cnt += 1
        if (cnt === N) {
            ans += 1
            cnt -= 1
        }
        i += 2
        
    } else {
        i += 1
        cnt = 0
    }

}

console.log(ans)